lib/Sema/CSDiagnostics.cpp

```diff
@@ -1227,40 +1259,63 @@ void AssignmentFailure::fixItChangeInoutArgType(const Expr *arg,
       .fixItReplaceChars(startLoc, endLoc, scratch);
 }
 
-std::pair<Expr *, ValueDecl *>
+std::pair<Expr *, Optional<OverloadChoice>>
 AssignmentFailure::resolveImmutableBase(Expr *expr) const {
   auto &cs = getConstraintSystem();
   auto *DC = getDC();
   expr = expr->getValueProvidingExpr();
 
+  auto isImmutable = [&](ValueDecl *decl) {
+    if (auto *storage = dyn_cast<AbstractStorageDecl>(decl))
+      return !storage->isSettable(nullptr) ||
+             !storage->isSetterAccessibleFrom(DC);
+
+    return false;
+  };
+
   // Provide specific diagnostics for assignment to subscripts whose base expr
   // is known to be an rvalue.
   if (auto *SE = dyn_cast<SubscriptExpr>(expr)) {
+
     // If we found a decl for the subscript, check to see if it is a set-only
     // subscript decl.
-    SubscriptDecl *member = nullptr;
-    if (SE->hasDecl())
-      member = dyn_cast_or_null<SubscriptDecl>(SE->getDecl().getDecl());
-
-    if (!member) {
-      auto loc =
-          cs.getConstraintLocator(SE, ConstraintLocator::SubscriptMember);
-      member = dyn_cast_or_null<SubscriptDecl>(getMemberRef(loc));
+    if (SE->hasDecl()) {
+      const auto &declRef = SE->getDecl();
+      if (auto *subscript =
+              dyn_cast_or_null<SubscriptDecl>(declRef.getDecl())) {
+        if (isImmutable(subscript))
+          return {expr, OverloadChoice(getType(SE->getBase()), subscript,
+                                       FunctionRefKind::DoubleApply)};
+      }
     }
 
+    Optional<OverloadChoice> member;
+
+    auto loc = cs.getConstraintLocator(SE, ConstraintLocator::SubscriptMember);
+    member = getMemberRef(loc);
```

Comment:
Can inline this locator call into `getMemberRef` since it's not going to be used anywhere else.